File: solana/scripts/omniswap/instructions/register_foreign_contract.py
from __future__ import annotations
import typing
from solders.pubkey import Pubkey
from solders.system_program import ID as SYS_PROGRAM_ID
from solders.sysvar import CLOCK
from solders.instruction import Instruction, AccountMeta
from anchorpy.borsh_extension import BorshPubkey
import borsh_construct as borsh
from ..program_id import PROGRAM_ID


class RegisterForeignContractArgs(typing.TypedDict):
    chain: int
    address: list[int]
    normalized_dst_base_gas_le: list[int]
    normalized_dst_gas_per_bytes_le: list[int]
    price_manager_owner: Pubkey
    init_price_ratio: int


layout = borsh.CStruct(
    "chain" / borsh.U16,
    "address" / borsh.U8[32],
    "normalized_dst_base_gas_le" / borsh.U8[32],
    "normalized_dst_gas_per_bytes_le" / borsh.U8[32],
    "price_manager_owner" / BorshPubkey,
    "init_price_ratio" / borsh.U64,
)


class RegisterForeignContractAccounts(typing.TypedDict):
    owner: Pubkey
    config: Pubkey
    foreign_contract: Pubkey
    price_manager: Pubkey
    token_bridge_foreign_endpoint: Pubkey
    token_bridge_program: Pubkey


def register_foreign_contract(
    args: RegisterForeignContractArgs,
    accounts: RegisterForeignContractAccounts,
    program_id: Pubkey = PROGRAM_ID,
    remaining_accounts: typing.Optional[typing.List[AccountMeta]] = None,
) -> Instruction:
    keys: list[AccountMeta] = [
        AccountMeta(pubkey=accounts["owner"], is_signer=True, is_writable=True),
        AccountMeta(pubkey=accounts["config"], is_signer=False, is_writable=False),
        AccountMeta(
            pubkey=accounts["foreign_contract"], is_signer=False, is_writable=True
        ),
        AccountMeta(
            pubkey=accounts["price_manager"], is_signer=False, is_writable=True
        ),
        AccountMeta(
            pubkey=accounts["token_bridge_foreign_endpoint"],
            is_signer=False,
            is_writable=False,
        ),
        AccountMeta(
            pubkey=accounts["token_bridge_program"], is_signer=False, is_writable=False
        ),
        AccountMeta(pubkey=SYS_PROGRAM_ID, is_signer=False, is_writable=False),
        AccountMeta(pubkey=CLOCK, is_signer=False, is_writable=False),
    ]
    if remaining_accounts is not None:
        keys += remaining_accounts
    identifier = b"\x84t\xa6\xf8Cm\x08\xe3"
    encoded_args = layout.build(
        {
            "chain": args["chain"],
            "address": args["address"],
            "normalized_dst_base_gas_le": args["normalized_dst_base_gas_le"],
            "normalized_dst_gas_per_bytes_le": args["normalized_dst_gas_per_bytes_le"],
            "price_manager_owner": args["price_manager_owner"],
            "init_price_ratio": args["init_price_ratio"],
        }
    )
    data = identifier + encoded_args
    return Instruction(program_id, data, keys)

File: solana/programs/omniswap/src/instructions/register_foreign_contract.rs
use anchor_lang::prelude::*;
use wormhole_anchor_sdk::{token_bridge, wormhole};

use crate::{
	error::SoSwapError,
	state::{ForeignContract, PriceManager, SenderConfig},
};

#[derive(Accounts)]
#[instruction(chain: u16)]
pub struct RegisterForeignContract<'info> {
	#[account(mut)]
	/// Owner of the program set in the [`SenderConfig`] account. Signer for
	/// creating [`ForeignContract`] account.
	pub owner: Signer<'info>,

	#[account(
		has_one = owner @ SoSwapError::OwnerOnly,
		seeds = [SenderConfig::SEED_PREFIX],
		bump
	)]
	/// Sender Config account. This program requires that the `owner` specified
	/// in the context equals the pubkey specified in this account. Read-only.
	pub config: Box<Account<'info, SenderConfig>>,

	#[account(
		init_if_needed,
		payer = owner,
		seeds = [
			ForeignContract::SEED_PREFIX,
			&chain.to_le_bytes()[..]
		],
		bump,
		space = ForeignContract::MAXIMUM_SIZE
	)]
	/// Foreign Contract account. Create this account if an emitter has not been
	/// registered yet for this Wormhole chain ID. If there already is a
	/// contract address saved in this account, overwrite it.
	pub foreign_contract: Box<Account<'info, ForeignContract>>,

	#[account(
		init_if_needed,
		payer = owner,
		seeds = [
			ForeignContract::SEED_PREFIX,
			&chain.to_le_bytes()[..],
			PriceManager::SEED_PREFIX
		],
		bump,
		space = PriceManager::MAXIMUM_SIZE
	)]
	/// Price manager account. Create this account if an emitter has not been
	/// registered yet for this Wormhole chain ID. If there already is a
	/// contract address saved in this account, overwrite it.
	pub price_manager: Account<'info, PriceManager>,

	#[account(
		seeds = [
			&chain.to_be_bytes(),
			token_bridge_foreign_endpoint.emitter_address.as_ref()
		],
		bump,
		seeds::program = token_bridge_program
	)]
	/// Token Bridge foreign endpoint. This account should really be one
	/// endpoint per chain, but Token Bridge's PDA allows for multiple
	/// endpoints for each chain. We store the proper endpoint for the
	/// emitter chain.
	pub token_bridge_foreign_endpoint: Account<'info, token_bridge::EndpointRegistration>,

	/// Token Bridge program.
	pub token_bridge_program: Program<'info, token_bridge::program::TokenBridge>,

	/// System program.
	pub system_program: Program<'info, System>,

	/// Clock used for price manager.
	pub clock: Sysvar<'info, Clock>,
}

pub fn handler(
	ctx: Context<RegisterForeignContract>,
	chain: u16,
	address: [u8; 32],
	normalized_dst_base_gas_le: [u8; 32],
	normalized_dst_gas_per_bytes_le: [u8; 32],
	price_manager_owner: Pubkey,
	init_price_ratio: u64,
) -> Result<()> {
	// Foreign emitter cannot share the same Wormhole Chain ID as the
	// Solana Wormhole program's. And cannot register a zero address.
	require!(
		chain > 0 && chain != wormhole::CHAIN_ID_SOLANA && !address.iter().all(|&x| x == 0),
		SoSwapError::InvalidForeignContract,
	);

	// Save the emitter info into the ForeignEmitter account.
	let emitter = &mut ctx.accounts.foreign_contract;
	emitter.chain = chain;
	emitter.address = address;
	emitter.token_bridge_foreign_endpoint = ctx.accounts.token_bridge_foreign_endpoint.key();
	emitter.normalized_dst_base_gas = normalized_dst_base_gas_le;
	emitter.normalized_dst_gas_per_bytes = normalized_dst_gas_per_bytes_le;

	let price_manager = &mut ctx.accounts.price_manager;
	price_manager.owner = price_manager_owner;
	price_manager.current_price_ratio = init_price_ratio;
	price_manager.last_update_timestamp = ctx.accounts.clock.unix_timestamp as u64;

	// Done.
	Ok(())
}
